cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/AppMetadataStoreTest.java

```diff
@@ -133,57 +136,193 @@ public void apply() throws Exception {
     // Add some run records
-    TreeSet<Long> expected = new TreeSet<>();
     ApplicationId application = NamespaceId.DEFAULT.app("app");
     final ProgramId program = application.program(ProgramType.WORKFLOW, "program");
-    final RunId runId = RunIds.generate(10000);
-    // Program completed status is not recorded with a smaller sourceId
+    // use sourceId's to ensure runId is unique for each run
+    final RunId runId = RunIds.generate(10000 + startSourceId + runningSourceId + killedSourceId);
+    final AtomicReference<RunRecordMeta> resultRecord = new AtomicReference<>();
     txnl.execute(new TransactionExecutor.Subroutine() {
       @Override
       public void apply() throws Exception {
         metadataStoreDataset.recordProgramStart(program, runId.getId(), RunIds.getTime(runId, TimeUnit.SECONDS),
                                                 null, ImmutableMap.<String, String>of(),
                                                 ImmutableMap.<String, String>of(),
-                                                AppFabricTestHelper.createSourceId(10000));
+                                                AppFabricTestHelper.createSourceId(startSourceId));
+        metadataStoreDataset.recordProgramRunning(program, runId.getId(), RunIds.getTime(runId, TimeUnit.SECONDS),
+                                                  null, AppFabricTestHelper.createSourceId(runningSourceId));
         metadataStoreDataset.recordProgramStop(program, runId.getId(), RunIds.getTime(runId, TimeUnit.SECONDS),
                                                ProgramRunStatus.KILLED,
-                                               null, AppFabricTestHelper.createSourceId(10));
+                                               null, AppFabricTestHelper.createSourceId(killedSourceId));
+        resultRecord.set(metadataStoreDataset.getRun(program, runId.getId()));
       }
     });
+    Assert.assertEquals(expectedRunStatus, resultRecord.get().getStatus());
+  }
 
-    final List<RunRecordMeta> runRecordMetas = new ArrayList<>();
+  private void assertPersistedStatus(final AppMetadataStore metadataStoreDataset, TransactionExecutor txnl,
+                                     @Nullable ProgramRunStatus expectedRunStatus,
+                                     List<ProgramRunStatus> recordStatuses)
+    throws Exception {
+    // Add some run records
+    ApplicationId application = NamespaceId.DEFAULT.app("app");
+    final ProgramId program = application.program(ProgramType.WORKFLOW, "program");
+    final RunId runId = RunIds.generate(10000);
+    final AtomicLong sourceId = new AtomicLong();
+    final Map<ProgramRunStatus, Runnable> statusRecordMap =
+      getStatusRecordMap(metadataStoreDataset, program, runId, sourceId);
+    // Get the record status runnable for each program run status
+    final List<Runnable> recordRunnables = Lists.transform(recordStatuses, new Function<ProgramRunStatus, Runnable>() {
```

Comment:
All this is indirection is overly complex and turns what should be a simple test case into something complicated. Remove this method and just record program status directly on the AppMetadataStore. It's much easier to understand and maintain if you just make a bunch of calls to recordProgramStart, recordProgramRunning, etc instead of changing a list of ProgramRunStatus into a list of Runnable through a map across three different methods.